Fixes #15634 - RancherOS support

zozoh94 opened the pull request:
I improved Semilson template to fix his own pull request : [](https://github.com/theforeman/community-templates/pull/450)

Commit:
Fixes #15634 - RancherOS support

lzap commented on partition_tables_templates/rancheros_default_fake.erb:
```diff
@@ -0,0 +1,8 @@
+oses:
+- RancherOS
+%>
+# Not required for RancherOS provisioning.
```
This might be useful for the future, can you rename the file and rename it to simply "Empty". Then keep the RancherOS association, but do not put any text, it must render to empty string. To make sure there is not even a newline add the minus sign in the end:

```erb
<%#
kind: ptable
name: Empty
model: Ptable
oses:
 - RancherOS
-%>
```

zozoh94 replied:
In a "empty.erb" file ?

lzap commented on provisioning_templates/PXELinux/rancheros_pxelinux.erb:
```diff
@@ -0,0 +1,12 @@
+
+LABEL rancheros
+  KERNEL <%= @kernel -%> rancher.password='changeme' rancher.state.dev=LABEL=RANCHER_STATE rancher.state.autoformat=[/dev/xvda] rancher.cloud_init.datasources=['url:<%= foreman_url('provision')-%>']
+  APPEND initrd=<%= @initrd -%>
```
Foreman can provide root password via `<%= root_pass %>` but it will be in Linux crypt form (`$5$xxxx$yyyyyyy`), not sure if RancherOS accepts it in this form. Putting password via plaintext is something you do not want to do, this file is word-readable via TFTP! I suggest to delete whole `rancher.password='changeme'` part if the OS boots without it.

Device should be parametrized to make this template usable, your will only work for Xen, use this: `<%= host_param('autoformat_device') %>` or similar.

zozoh94 replied:
Rancher don't support crypted passwor I tried. So I will delete this parameter. I will be the responsability of the user to either have public keys confgured or add this parameter.

lzap replied:
Thanks, since it supports cloud-init this is not an issue actually.

lzap replied:
The line is a bit off now, can you add two leading spaces?

This would work as well: `<%= host_param('autoformat_device') || '/dev/vda' %>`

mmoll replied:
The `install-disk` parameter should be re-used here, IMHO

lzap commented on provisioning_templates/snippet/rancheros_cloudconfig.erb:
```diff
@@ -0,0 +1,52 @@
+#cloud-config
```
Please put this line below the comment, it will work.

lzap reviewed (changes requested):
Couple of changes to have better default templates.

lzap commented on provisioning_templates/PXELinux/rancheros_pxelinux.erb:
```diff
@@ -0,0 +1,12 @@
+DEFAULT rancheros
+
+LABEL rancheros
+  KERNEL <%= @kernel -%> rancher.state.dev=LABEL=RANCHER_STATE rancher.state.autoformat=[/dev/xvda] rancher.cloud_init.datasources=['url:<%= foreman_url('provision')-%>']
```
The `/dev/xvda` is still unparametrized, can you add a param and default to this when not set?

mmoll replied:
The `install-disk` parameter should be re-used here, IMHO

zozoh94 replied:
So install-disk or autoformat_device ?

mmoll replied:
install-disk please, that variable is used in some other templates.

lzap reviewed (changes requested):
Last comment about the hardcoded device name and we are good to go!

mmoll commented:
thanks, I try to test this over the weekend :+1:

lzap reviewed (approved):
Thanks I am not going to test this end to end, so if you don't get to it let's just merge.

zozoh94 commented:
Wait before merging I will do some last tests to see if cloud-config work well.

lzap commented:
Thanks appreciated, let us know how it went. Preferably squash all commits into one before testing, so we all have the same SHA value.

zozoh94 commented:
I merged commits in just one by rebasing called : Fixes #15634 - RancherOS support. I tried to put rancher password in cloud-config file a last time but unfortunately it seems impossible but everything works well with authorized_keys. I also figure out that there is no binary for wget inside RancherOS for template finishing but there is a curl one so I remplace the long wget command with "curl -sk". I also added a default autoformat_device which I forgot when I replaced autoformat_device param with install-disk.

So I think it's ok you can merge !

zozoh94 commented:
Wait the finish template seems not working. I will find a fix.

zozoh94 commented:
It's ok I delete the finish template because Rancher work with a provisioning each time there is a reboot. Rancher works on RAM and will just mount the device marked with the label RANCHER_STATE to keep docker images and persist cloud-config file.

mmoll commented:
superseded by GH-464

mmoll commented:
this is for sure not perfect, e.g. that eth* and eth0 is hardcoded, but it seems to work, so let's get this in and improve later :+1:

mmoll commented:
merged, thanks @zozoh94 and @selmison!
